### PR TITLE
Add Duplicate Lines shortcut to CodeTextEditor

### PR DESCRIPTION
**Pull request by PucklaJ**
This PR adds the shortcut `script_text_editor/duplicate_lines`. The default buttons are `Ctrl+Alt+Down`. For macOS the default buttons are `Cmd+Shift+Down`. These shortcuts are similiar to the VS Code command `editor.action.copyLinesDownAction`. If no text is selected it behaves the same as `duplicate_selection`, but if text is selected it will duplicate the whole lines of the selection.
This also supports multiple carets.

`script_text_editor/duplicate_lines`:
duplicate_lines_below

`script_text_editor/duplicate_selection`:
duplicate_selection

### CODE REVIEW

**Comment by PucklaMotzer09** [comment]
Added suport for multiple carets

**Comment by MewPurPur** [comment]
I'll do a code review in a sec, but can you argue for why you think this feature should be added? We don't accept features just because PRs exist for them.

I personally think Ctrl+Shift+D does what I want just fine. If I have no selections, this PR does the same, just adjusts the caret in one of the scenarios. If I want to duplicate multiple lines, I'm just going to do a proper selection with triple-clicking and then dragging, both very well-known functionalities.

File: editor/code_editor.cpp
Comment (MewPurPur) [inline editor/code_editor.cpp:1702]: Likewise, just put this stuff in the `has_selection()` if/else.

File: editor/code_editor.cpp
Comment (MewPurPur) [inline editor/code_editor.cpp:1635]: You don't use these much outside of places where `has_selection()` is checked. You should define them close to the scope where the variables have a meaning.

**Review by MewPurPur — CHANGES_REQUESTED** [review]
Haven't tested the feature itself, just a code review.

**Comment by torcado194** [comment]
I would love this feature. I have never once wanted to duplicate a subsection within a line using a hotkey, *especially* if that selection spans multiple lines, I couldn't imagine a scenario in which I would want to duplicate that content in-place.
I duplicate lines constantly, duplicating and swapping lines are probably my most used hotkeys aside from undo/redo. Needing to make sure my selections encompass the entirety of the lines I want to duplicate before i'm able to duplicate multiple lines is very tedious, especially with no way to automatically spread selections to the start/end of lines. If a cursor in my multi-cursor selection started in the middle of the line? Now i have to redo the whole selection to move the start point to the beginning of the line.
And a mouse-based alternative isn't ideal. I'm already typing, I want to use the keyboard, not the mouse. At least for something this simple.

**Comment by MewPurPur** [comment]
@torcado194 That makes sense, I think "Duplicate Lines" is a hard thing to argue against. What about this particular implementation, with separate "below" and "above" versions? That's the main part I find overly complicated.

**Comment by torcado194** [comment]
I agree, I hardly ever use "duplicate above". "below" is the default direction for the duplicate line action for any editor I use (and is what I'd expect). the "duplicate above" result can be easily substituted using the keyboard by duplicating below and swapping the new selected lines upward, at the cost of additional keystrokes equal to the line count.
but on the other hand the duplicate above and below functions are found in many editors I've used (visual studio and vs code for example)
maybe it's not part of godot's philosophy but I would *much* rather have more hotkey functions than less, even if they're unassigned by default.

**Comment by PucklaJ** [comment]
> What about this particular implementation, with separate "below" and "above" versions?

I tried to recreate the feature from VS Code and there they have both a above and below version. But I agree I hardly use it either.

File: editor/code_editor.cpp
Comment (PucklaJ) [reply editor/code_editor.cpp:1637]: I tried to remove them, but when the caret moves, the selection gets unset, which means that I need to store them before changing the caret. And I think that I should transfer the selection to the duplication since it means that you can perform the same duplication multiple times in a row. Which is something I use very often.

File: editor/code_editor.cpp
Comment (PucklaJ) [reply editor/code_editor.cpp:1702]: I unfortunately can not put this before inserting the text, because I need to deselect when inserting text, because the inserted text replaces the selection.

File: editor/code_editor.cpp
Comment (PucklaJ) [reply editor/code_editor.cpp:1635]: If I put them in their own scope I would need to write duplicate code.

**Comment by PucklaJ** [comment]
I will merge all commits into one, once you are done reviewing

File: editor/code_editor.cpp
Comment (PucklaJ) [reply editor/code_editor.cpp]: If I could use `_insert_text`, which is private, I could put this before the text insertion, because it wouldn't affect the selection.

**Comment by PucklaJ** [comment]
@MewPurPur @Paulb23 This PR has just been hanging here for quite some time. Can you please finish reviewing it. I would really like it to be merged. I think it adds some good keyboard shortcuts which can really improve your productivity.

**Comment by MewPurPur** [comment]
I still think it should be just "Duplicate Lines" that does it below, instead of above and below versions. It's not like regular duplicate is "before" and "after" I don't got much else to say, the code to me looks good and my remarks have been clarified or addressed.

**Comment by PucklaJ** [comment]
Since it has been a few months since I last visited this PR I performed a rebase. On top of that I removed the duplicate_lines_above shortcut and only the below one stays with the short name duplicate_lines. Since duplicate_lines behaves the same as duplicate_selection without a selection I replaced it with duplicate_lines. In case you don't like this change I can just remove the last commit.
But I think that it makes sense to replace duplicate_selection, because I think that probably nobody has been using duplicate_selection with a selection. By replacing the old shortcut people would use the new one and notice the improvement.

**Comment by MewPurPur** [comment]
All in all, I agree, I think this does great with preventing bloat. I've misused duplicate selection a few times and never found usecases for it. Other opinions would be great to gather, as I don't use IDEs and Godot's script editor is the only thing I have lots of experience with.

**Comment by PucklaJ** [comment]
@Paulb23 What do you think?

File: editor/code_editor.cpp
Comment (MewPurPur) [inline editor/code_editor.cpp:1453]: `new_caret_line = select_to_line + select_num_lines;`

**Review by MewPurPur — APPROVED** [review]
Looked at it again - no viruses have been added, the code is better, I couldn't think of anything missing from it. As already mentioned, I think only the change itself could possibly rustle some jimmies.

File: editor/code_editor.cpp
Comment (PucklaJ) [reply editor/code_editor.cpp]: You are right. This has been overlooked since removing the duplicate_lines_above code.

**Comment by PucklaJ** [comment]
@Paulb23 @YuriSizov Could someone please also have a look at this PR and maybe even approve it? The PR has already deemed fine by @MewPurPur and the code itself is also okay. Could we please move on with this PR. It is just a key binding that would probably be a great addition for some users. Those you don't need or don't care about keybindings can just ignore it without any implications. I think there really is nothing to discuss anymore besides it replacing the `duplicate_selection` shortcut. Which I can just easily undo if this is an issue. Therefore I ask the Godot team to approve this PR and merge it.

**Comment by YuriSizov** [comment]
Paul is the area maintainer, so I would still prefer his approval on this rather than doing it myself. There is also no proposal for this, so it's unclear to me if this is something that would benefit many users, or just a few. 

That said, before this can be merged you need to squash your commits.

**Comment by PucklaJ** [comment]
> That said, before this can be merged you need to squash your commits.

Done

**Review by Paulb23 — COMMENTED** [review]
Code looks fine, few minor comments.

I'm wanting to move things like this into `CodeEdit`, could then take advantage of unit tests as well. If you fancy doing that :)

**Comment by PucklaJ** [comment]
@Paulb23 I moved the code into `CodeEdit` and then adjusted the code based on your review.

File: scene/gui/code_edit.h
Comment (Paulb23) [inline scene/gui/code_edit.h:468]: `nit:` Can title this with a new section i.e from `TextEdit`:  `// Text manipulation` or `// Input actions`

**Review by Paulb23 — COMMENTED** [review]
Thanks!

One small change, could you bind it to to the `CodeEdit` API? and again would appreciate any unit tests.

**Comment by PucklaJ** [comment]
@Paulb23 After a rebase and fixing a merge conflict I added unit tests

File: scene/gui/code_edit.cpp
Comment (Paulb23) [inline scene/gui/code_edit.cpp:2416]: Seems to be from #81551, but could you move this above `/* Visual */` so it's the same order as the header?

Otherwise looks great, really appreciate the unit tests too!

Comment (PucklaJ) [reply scene/gui/code_edit.cpp:2403]: Done

File: doc/classes/CodeEdit.xml
Comment (AThousandShips) [inline doc/classes/CodeEdit.xml:158]:
```suggestion
				Duplicates all lines currently selected with any caret. Duplicates the entire line beneath the current one no matter where the caret is within the line.
```
I believe more appropriate word

File: editor/plugins/script_text_editor.cpp
Comment (YuriSizov) [inline editor/plugins/script_text_editor.cpp:2403]: Is this a common macOS shortcut for this purpose?

**Review by YuriSizov — APPROVED** [review]
Consensus seems to be for this, at a glance the style looks fine.

**Comment by YuriSizov** [comment]
Could you please squash your commits into one? Make sure that the final commit has a short but descriptive message (the title of this PR is a good option). See [this documentation](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), if you need help with squashing.

File: editor/plugins/script_text_editor.cpp
Comment (PucklaJ) [reply editor/plugins/script_text_editor.cpp:2403]: I don't know. Honestly I just put anything there because it would have been a duplicate otherwise

**Comment by PucklaJ** [comment]
@YuriSizov @Paulb23 @MewPurPur The commits are squashed. Everything is ready. Thank you for approving this PR.

File: editor/plugins/script_text_editor.cpp
Comment (akien-mga) [reply editor/plugins/script_text_editor.cpp:2403]: Would be good to know what's the expected macOS shortcut for this operation (e.g. in VS Code or Xcode). CC @bruvzg @coppolaemilio

Comment (PucklaJ) [reply editor/plugins/script_text_editor.cpp:2403]: I just [looked it up](https://code.visualstudio.com/shortcuts/keyboard-shortcuts-macos.pdf). On VS Code its `shift+cmd+down`

Comment (akien-mga) [reply editor/plugins/script_text_editor.cpp:2403]: That might work. Seems like we use this mapping but only for PC here, and there's a different override for macOS:
```cpp
// core/input/input_map.cpp
        inputs = List<Ref<InputEvent>>();
        inputs.push_back(InputEventKey::create_reference(Key::DOWN | KeyModifierMask::SHIFT | KeyModifierMask::CMD_OR_CTRL));
        default_builtin_cache.insert("ui_text_caret_add_below", inputs);

        inputs = List<Ref<InputEvent>>();
        inputs.push_back(InputEventKey::create_reference(Key::L | KeyModifierMask::SHIFT | KeyModifierMask::CMD_OR_CTRL));
        default_builtin_cache.insert("ui_text_caret_add_below.macos", inputs);
```

Comment (PucklaJ) [reply editor/plugins/script_text_editor.cpp]: I changed it on macos to `cmd_or_ctrl+shift+down`

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
